core/src/main/java/io/undertow/server/handlers/CurlRequestDumpingConduit.java

```diff
@@ -0,0 +1,80 @@
+package io.undertow.server.handlers;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.xnio.IoUtils;
+import org.xnio.channels.StreamSinkChannel;
+import org.xnio.conduits.AbstractStreamSourceConduit;
+import org.xnio.conduits.ConduitReadableByteChannel;
+import org.xnio.conduits.StreamSourceConduit;
+
+public class CurlRequestDumpingConduit extends AbstractStreamSourceConduit<StreamSourceConduit>
+{
+
+    private final List<byte[]> data = new CopyOnWriteArrayList<>();

```

Comment:
Since this is created per request I don't think it needs to be thread safe. A normal list should be fine.